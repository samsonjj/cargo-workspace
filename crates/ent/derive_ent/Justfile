test:
    cargo test --lib -- --nocapture