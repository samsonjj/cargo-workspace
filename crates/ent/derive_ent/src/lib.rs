use convert_case::{Case, Casing};
use proc_macro2::TokenStream;
use quote::quote;
use syn::ItemStruct;

mod derive_fields;

type Ast = syn::ItemStruct;

fn parse(tokens: proc_macro2::TokenStream) -> Ast {
    let ast: ItemStruct = syn::parse2(tokens.into()).unwrap();
    ast
}

#[derive(Debug, PartialEq)]
enum ModelFieldType {
    Int,
    String,
}

#[derive(Debug, PartialEq)]
struct Model {
    fields: Vec<ModelField>,
}

#[derive(Debug, PartialEq)]
struct ModelField {
    ident_rust: String,
    ty: ModelFieldType,
}

fn analyze(ast: Ast) -> Model {
    let mut model = Model { fields: vec![] };

    // collect fields
    model.fields = ast
        .fields
        .iter()
        .map(|f| {
            let ident_rust = f.ident.as_ref().unwrap().to_string();
            let ty = match &f.ty {
                syn::Type::Path(tp) => {
                    assert!(
                        tp.path.segments.len() == 1,
                        "type path segments has a length that is not 1"
                    );
                    println!("hello");
                    match tp.path.segments[0].ident.to_string().as_ref() {
                        "i32" => ModelFieldType::Int,
                        "String" => ModelFieldType::String,
                        id => panic!("Unsupported field type {:?}", id),
                    }
                }
                x => panic!("Unsupported field type {:?}", quote!(#x)),
            };
            ModelField { ident_rust, ty }
        })
        .collect();

    model
}

#[proc_macro_derive(Ent, attributes(primary_key))]
pub fn derive_ent(tokens: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let ast: Ast = parse(tokens.into());
    let model = analyze(ast);
    TokenStream::new().into()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse() {
        let tokens = quote!(
            struct Point {
                x: i32,
                y: i32,
            }
        );
    }

    #[test]
    fn ui() {
        let t = trybuild::TestCases::new();
        t.compile_fail("tests/ui/*.rs");
    }

    #[test]
    fn test_analyze() {
        let tokens = quote!(
            struct Foo {
                bar: i32,
                message: String,
            }
        );

        let model = analyze(parse(tokens));

        assert_eq!(
            model,
            Model {
                fields: vec![
                    ModelField {
                        ident_rust: "bar".to_string(),
                        ty: ModelFieldType::Int,
                    },
                    ModelField {
                        ident_rust: "message".to_string(),
                        ty: ModelFieldType::String,
                    }
                ]
            }
        );
    }
}
