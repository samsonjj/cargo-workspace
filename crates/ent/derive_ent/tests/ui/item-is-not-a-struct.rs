use derive_ent::Ent;

// you will get rust-analyzer errors here, it's intentional :)
#[derive(Ent)]
enum Foo {}

fn main() {}
