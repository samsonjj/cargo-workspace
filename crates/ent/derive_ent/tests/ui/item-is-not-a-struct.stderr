error: proc-macro derive panicked
 --> tests/ui/item-is-not-a-struct.rs:4:10
  |
4 | #[derive(Ent)]
  |          ^^^
  |
  = help: message: called `Result::unwrap()` on an `Err` value: Error("expected `struct`")
